### ISSUE TITLE
mutable coordinates

### ISSUE THREAD

**Issue by hainm** [ParmEd/ParmEd#302]
@swails 

I think parm should warn user about trying to mutate `coordinates`

```python
In [14]: p.coordinates[0, 0]
Out[14]: 16.55458927154541

In [15]: p.coordinates[0, 0] = 17.

In [16]: p.coordinates[0, 0]
Out[16]: 16.55458927154541
```

`pytraj` has the same problem, so I need to write a immutable [XYZ array](https://github.com/pytraj/pytraj/blob/master/pytraj/_xyz.py)
(but this is kind of annoying, so I am thinking about switch `Trajectory` class to the one like `mdtraj`, which completely use numpy array)

**Comment by swails** [comment]
So it's not really immutable.  It's just that `coordinates` is a second-class citizen of sorts.  Each atom has a `xx`, `xy`, and `xz` attribute, and the policy is that these are *always* correct.  So in your example, you *did* change `p.coordinates`, but when you try to access the descriptor again, ParmEd checks against each atom's existing coordinates, finds that they're different, and discards the cached `_coordinates` numpy array and regenerates it with the original ones.

Now, *assigning* an entire `coordinates array` is different -- in that case, those coordinates updated to the `Atom`s in the structure.  So for example, the following should work:

```Python
>>> coords = p.coordinates
>>> coords[0,0]
16.55458927154541
>>> coords[0,1] = 17.0
>>> p.coordinates = coords
>>> p.coordinates[0, 0]
17
>>> # And just for good measure:
... p.atoms[0].xx
17
```

I don't know how I would change your syntax into something that would update the Atom coordinates (or I would do that in a second :)).  Nor do I know how to easily make the coordinates numpy array immutable without making it no longer a numpy array.

I realize this is surprising behavior, but I'm not exactly sure how to fix it.  Ideas welcome.

**Comment by hainm** [comment]
No, this is not surprising. It is simply because you make a copy of atoms coordinates whenever calling p.coordinates. 

This happens since list of atoms does not have contiguous memory. numpy needs to copy to make contiguous one.

Here is two options
1. You can wrap numpy array like I did in pytraj (link in my 1st post) that turn off __setitem__

To make it fully numpy array, you can use 'p.coordinates[:]' (this is what pytraj does now).

2. Simply add to doc, explicitly saying that user should not __setitem__

Options 2 is preferred 

PS: 'p.coordinates = coords' this example works because you update all coords. Which is very expensive to large system if user just want to update coords for only few atoms.

**Comment by hainm** [comment]
Look closer here. Feel free to copy this code if you want to implement it in parmed.

https://github.com/pytraj/pytraj/blob/master/pytraj/_xyz.py

**Comment by swails** [comment]
> No, this is not surprising. It is simply because you make a copy of atoms coordinates whenever calling p.coordinates.

Agreed -- it's not surprising if you know the implementation ;).  If you don't (and don't know the coordinate precedence rules in ParmEd), then it *is* surprising, I think.  What I'd ideally like is a way to have

```Python
p.coordinates[0,0] = 17.0
```

automatically update `p.atoms[0].xx`.  I just don't know how to do that without deeply convoluted code.

> 1. You can wrap numpy array like I did in pytraj (link in my 1st post) that turn off __setitem__

This then fails typechecks and will likely fail when used in a variety of places in the scipy stack.  Also, what happens if you do something like `p.coordinates[:][0,0] = 17.0`?  These kinds of simple slices in numpy returns views, so one should expect the following type of behavior:

```Python
>>> import numpy as np
>>> x = np.arange(102).reshape((-1, 3))
>>> x.shape
(34, 3)
>>> x[0]
array([0, 1, 2])
>>> x[:][0,0] = 10
>>> x[0]
array([10,  1,  2])
```

Which ParmEd would clearly violate.

**Comment by hainm** [comment]
Agree. 

So it's better just to warn user about this.  I don't think we can do anything else.

**Comment by swails** [comment]
I'm not even sure how to warn this behavior unless it's just in the docs...

**Comment by hainm** [comment]
Yeah, just in doc.

On Saturday, July 25, 2015, Jason Swails <notifications@github.com> wrote:

> I'm not even sure how to warn this behavior unless it's just in the docs...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ParmEd/ParmEd/issues/302#issuecomment-124933075>.
>


-- 
Hai Nguyen

**Comment by swails** [comment]
Fixed in #350
